### DISCUSSION TITLE
Von SD-card zu flash

### DISCUSSION THREAD

**Discussion by Marvin1DE** [RudolphRiedel/FT800-FT813#110] [Q&A] [answered]
Hey Marvin nochmal hier, 
Ich habe etwas Probleme damit die "test.bin" bilddatein von meiner SD Karte auf den flash speicher zu schieben, kannst du oder jemand mir ein beispiel dazu geben, habe die SD Karte an SPID und den Grafikchip an SPIC, hab ein Filesystem das funktioniert auch kann drauf lesen & schreiben aber jedoch brauche ich mehrere Bilder & später UTF-8 Schriften die ich von der SD Karte auf den flash schieben sollte,
hab einen xmega384C3 mit dem EVE817q, 16 mb falsh speicher an dem BT817, ich werde nicht daraus schlau wie ich genau die bin datein von der SD-Karte auf den flash speicher bekomme und dann nutze.

ich bedanke mich schonmal für die zeit.

**Comment by RudolphRiedel** [comment, marked as answer]
Moin,

das letzte Mal das ich eine SD-Karte an einem Controller verwendet habe ist über 10 Jahre her -> gar keine Ahnung von.

Generell haben die Dinger ja Sekoren, ein Datei-System müsste eigentlich erlauben eine Datei Sektor-für-Sektor zu lesen.

Auf der EVE Seite hast Du CMD_FLASHWRITE, CMD_FLASHPROGRAM und CMD_FLASHUPDATE.
CMD_FLASHWRITE: direkt per SPI, erfordert vorheriges löschen, die Ziel-Adresse muss durch 256 teilbar sein und die Größe muss ein vielfaches von 256 haben
CMD_FLASHPROGRAM: aus dem RAM_G, erfordert vorheriges löschen, die Quell-Adresse muss durch 4 teilbar sein, die Ziel-Adresse muss durch 4096 teilbar sein und die Größe muss ein vielfaches von 4096 haben
CMD_FLASHUPDATE: aus dem RAM_G, löscht wenn erforderlich selber, die Quell-Adresse muss durch 4 teilbar sein, die Ziel-Adresse muss durch 4096 teilbar sein und die Größe muss ein vielfaches von 4096 haben

Also die erste Frage ist mal, wie groß ist ein Sektor der von der SD-Karte gelesen wird, bzw. vielleicht liest man einen Sektor ja auch nicht in einem Rutsch.

Sagen wir mal 256 Bytes und per CMD_FLASHUPDATE

CMD_MEMSET - erstmal 4096 Bytes irgendwo im RAM_G auf 0xff setzen
Sektor von SD lesen
Sektor an EVE schicken (EVE_memWrite_sram_buffer() )
wiederholen bis 4096 voll oder Datei zu Ende
CMD_FLASHUPDATE
wiederholen bis Datei zu Ende

So in etwa, da gibt es bestimmt noch ein paar gemeine Details drin. :-)

**Reply by Marvin1DE** [reply]
danke schonmal, wird eine weile dauern, auf der eine platine geht das filesystem ohne probleme, aber auf der anderen ist sie anders angeschlossen werde erstmal den schaltplan brauchen, jedenfalls danke für die schnelle antwort
